### ISSUE TITLE
Stream Byte Flow Control

### ISSUE THREAD

**Issue by benjchristensen** [ReactiveSocket/reactivesocket#83]
ReactiveSocket already support 2 types of flow control mechanisms:

- REQUEST_N for application level management of number of items being sent on a stream
- LEASE for a Responder to control the number of requests it can receive

There is a potential deficiency if large messages are sent that REQUEST_N will not necessary address.

For example, if a large image is sent, it could head-of-line block all other streams, since the image is considered a single item, but could be very large. 

An implementation can correctly use small fragment sizes in the protocol to ensure it interleaves correctly, but nothing about the protocol currently ensures this is in fact how an implementation will behave. Also, even with fragmentation, nothing will stop that stream from sending everything. 

So, should we consider modifying the protocol to support the following two improvements?

1) Default fragmentation size that is generally optimal for public internet (and can be changed if something is better suited in other environments)
2) Byte level credit based flow control like HTTP/2 (https://httpwg.github.io/specs/rfc7540.html#WINDOW_UPDATE) per stream

**Comment by NiteshKant** [comment]
I am on the fence for defaulting to a lower data size than frame size (currently max of ~2GB) for this reason as it would also be a function of number of concurrent streams. I think a sender can at runtime decide to optimize and fragment based on the load/concurrency on the connection. A global default (sort of a setting) restricts such optimizations with limited benefits IMHO.

Byte level flow control (windowing) is certainly useful but is complex and will mostly be hidden from an application as that will be providing demand via `request(n)` semantics. So, it will most likely be limited only to a transport implementation that I think makes it less useful in presence of a `request(n)` semantics.

I am inclined towards not adding these and leave it to the implementations to optimize using fragmentation as they perceive correct but I am interested in hearing what others have to say about this.

**Comment by tmontgomery** [comment]
Both end-to-end application data unit and byte-level flow control are needed. So, another option might be to always use a byte-level flow control transport protocol below. Which is provided by TCP and Aeron.

**Comment by NiteshKant** [comment]
@tmontgomery agreed, however, as I understand this is around per stream flow control (byte-level) as what HTTP/2 specifies with WINDOW_UPDATE and not for the entire connection as what TCP provides with windowing.

**Comment by benjchristensen** [comment]
Consider this example:

- REQUEST_STREAM(A) from client to server with N == 10
- server starts sending items that are 10MB each (fragmented however it wishes)
- client is slowly consuming items from A
- REQUEST_RESPONSE(B) from client to server
- server sends 1KB response for B

If fragments are small, B can be interleaved with A, but the stream from A will have saturated the network and filled all byte buffers along the path. The REQUEST_N semantics do not govern the fragments, only the application layer count of items. 

Thus, the standing queue in the network will be at the connection level, and defined by however fast the 100MB (10MB * 10) of A items are put on the network. 

Thus, the TCP/Aeron level flow control is only going to govern the connection level bytes, not the individual streams like HTTP/2 WINDOW_UPDATE. 

I suggest that we need to permit limiting the flow of bytes per stream, just like HTTP/2 does on top of TCP, so that in the example above, A does not saturate the network buffers.

**Comment by benjchristensen** [comment]
As the HTTP/2 spec states:

> Flow control operates at two levels: on each individual stream and on the entire connection.

https://httpwg.github.io/specs/rfc7540.html#WINDOW_UPDATE

and:

> Using streams for multiplexing introduces contention over use of the TCP connection, resulting in blocked streams. A flow-control scheme ensures that streams on the same connection do not destructively interfere with each other. Flow control is used for both individual streams and for the connection as a whole.

https://httpwg.github.io/specs/rfc7540.html#FlowControl

**Comment by benjchristensen** [comment]
> I am on the fence for defaulting to a lower data size than frame size (currently max of ~2GB) for this reason as it would also be a function of number of concurrent streams.

Why would the lower size affect anything negatively though? If there is no other contention, the fragments would flow without interleaving. The throughput should hardly be affected, other than the overhead of fragmenting and reassembling, and the bytes per frame, which are trivial compared with a large frame of bytes. 

> Byte level flow control (windowing) is certainly useful but is complex and will mostly be hidden from an application as that will be providing demand via request(n) semantics. 

Yes it is hidden from the app, purely for the protocol implementations, and that's a good thing. 

The intent is that if a stream is not being consumed at the byte level, it should only buffer that stream, and not block other multiplexed streams. 

If a consumer is reading from a single stream fast enough to keep up, then that single stream can saturate the entire connection. (For example, if only a single stream is being used over the connection).

**Comment by tmontgomery** [comment]
Congestion control, not flow control, is needed to not overflow the network buffering on a hop-by-hop basis. This level of clarity is not what HTTP/2 (or anything built on top of TCP) has and why TCP provides congestion control.

**Comment by benjchristensen** [comment]
I understand this is different from what TCP does, and I'm not claiming to reinvent congestion control. That's a whole different discussions for things like QUIC, Sprout, Giga, etc. 

I'm specifically talking about the contention between multiplexed streams, and the reason for why HTTP/2 has WINDOW_UPDATE.

**Comment by NiteshKant** [comment]
@benjchristensen from what you are describing, it sounds to me more a case for stream priorities as opposed to byte-level flow control. Let us take the case you describe:

```
REQUEST_STREAM(A) from client to server with N == 10
server starts sending items that are 10MB each (fragmented however it wishes)
client is slowly consuming items from A
REQUEST_RESPONSE(B) from client to server
server sends 1KB response for B
```

and let us assume that the transport does some fragmentation for larger payloads by heuristics local to the transport i.e. it does not send 10MB frames in this case. Unless, the window size is pretty low (HTTP/2 default is 2 gigs), the transport will not give a chance to other streams to write, unless it knows that the other streams are important. It may be that B is low priority data and A is high, in which case, if there is space to send, it should prefer sending A.

This is different than HTTP/2 as there, consumer of the data would typically govern the next request. OTOH, in our case, the consumer of data is governing `request(n)` so there isn't a meaningful way byte window sizes can be altered.

Now, if we take stream priorities in account, the scenario you described, is easier to solve by determining whether B is really important to send. If so, the transport should fragment and try to be fair w.r.t priorities.

**Comment by stevegury** [comment]
Fragmentation was designed to take care of multiplexing different streams of data together on the same TCP connection. The TCP flow control then take care of controlling the flow of bytes.
If my understanding of HTTP2 WINDOW_UPDATE is correct, it is mostly usefull for dealing with streams of different priorities.

That being said there is a case where WINDOW_UPDATE could be useful in reactive socket: If we want to garantee "bandwith fairness" between multiple connections, i.e. garantee that every client receives its fair share of the server bandwith, not having one client sucking it all because it fetches large message.

**Comment by tmontgomery** [comment]
@benjchristensen just clarifying your statements about network buffering. Nothing HTTP/2, or RS, does is for network buffering control.

I see what you are talking about. And I actually, partially, agree. But also, HTTP/2 is not the same as RS. 

I do agree that per-stream byte-level flow control is needed in addition to application-unit flow control. However, per-stream flow control (whether RS or HTTP/2) introduces a need for scheduling. That need also implies need for control of connection-level flow control. HTTP/2 must control that or have the right ability to control it at the transport level.

This is, IMO, an all-or-nothing approach. Either the protocol takes control of everything or it leaves some of that work up to the transport.

**Comment by benjchristensen** [comment]
The WINDOW_UPDATE and prioritization aspects of HTTP/2 are separate. An "elephant flow" on a single multiplexed stream can starve the other multiplexed streams. That's what I'm trying to solve. REQEUST_N and fragments don't prevent that. 

>  Either the protocol takes control of everything or it leaves some of that work up to the transport.

Then why does HTTP/2 need to have WINDOW_UPDATE on top of TCP? It's not for streaming which REQUEST_N is for, it's for handling images, video, html, etc as byte streams.

If RS didn't have streams, and was just REQUEST_RESPONSE, it wouldn't have REQUEST_N, and would need something like WINDOW_UPDATE, correct? 

Thus, RS, when used with just REQUEST_RESPONSE, likely needs the same.

**Comment by tmontgomery** [comment]
@benjchristensen I agree. HTTP/2 takes control of the connection flow control so that implementations can schedule frames effectively. It's a frame/fragment scheduling issue. Or at least, this is how I believe it should be addressed in HTTP/2 implementations. And that is very hard to do unless the connection is also modeled at that layer... or the transport gives the right control for the app to do it. Which TCP does not.

Yes, if RS just had REQUEST_RESPONSE, it should have some mechanism (at byte-level) to control per stream RESPONSE flow. Each stream should be flow controllable separately. And yes, that extends to each type of request when you add REQUEST_N.

RS does suffer from potential starvation. But then, and this is personal opinion, ReactiveStreams does also. When the object size (1 of N) is large, then memory and CPU can starve out other objects. This can be addressed, in a closed system, by making sure that object size is "reasonable" for the use case.

However, In a general use case, though, this would fall over as object size could be totally unbounded.

**Comment by tmontgomery** [comment]
Let us revive this. Especially in relation to resumability (#102).

**Comment by benjchristensen** [comment]
@tmontgomery Is there anything about this that we should address?

**Comment by benjchristensen** [comment]
Do these slides correctly represent the relationship of TCP, HTTP/2, and ReactiveSocket?

And if so, does ReactiveSocket actually need yet another?

![reactivesocket-dec2016 031](https://cloud.githubusercontent.com/assets/813492/21862012/502ce9ec-d7ec-11e6-965c-6af43e103300.jpeg)
![reactivesocket-dec2016 032](https://cloud.githubusercontent.com/assets/813492/21862014/504b4216-d7ec-11e6-97ea-0eed14b3ebc5.jpeg)
![reactivesocket-dec2016 033](https://cloud.githubusercontent.com/assets/813492/21862016/505059cc-d7ec-11e6-85b2-d3ab4a3a86f3.jpeg)
![reactivesocket-dec2016 034](https://cloud.githubusercontent.com/assets/813492/21862015/504c3c70-d7ec-11e6-8ce2-11c1c5f5649b.jpeg)
![reactivesocket-dec2016 035](https://cloud.githubusercontent.com/assets/813492/21862013/5048e7b4-d7ec-11e6-8067-7021fb250ed5.jpeg)

**Comment by tmontgomery** [comment]
I think the slides are accurate. And with HTTP/2, we don't need anything else. However, with TCP alone, there could be need for more.
